state/presence/presence.go

```diff
@@ -353,6 +353,11 @@ func (w *Watcher) flush() {
 				return
 			case req := <-w.request:
 				w.handle(req)
+				// handle may append to the w.pending array, and/or it may unwatch something that was previously pending
```

Comment:
ISTM that we're already handling the unwatch case (that's the `e.ch != nil` loop condition) but that the real issue here is the append, because the append can cause a realloc and thus e might be pointing to the old slice and hence be invalid.

I think I'd probably program this a little more defensively;

     for {
         // Note: w.pending can be appended to by w.handle so we can't
         // safely keep the same e between iterations.
         e := &w.pending[i]
         if e.ch == nil {
             break
         }
         select {
         etc

?